torchtune/_cli/tune.py

```diff
@@ -4,133 +4,49 @@
-          declared in the ``scripts`` configuration in
-          `setup.py <https://github.com/pytorch/torchtune/blob/main/setup.py>`_.
-          It is equivalent to invoking ``python -m scripts.cli_utils.tune``.
-"""
 import argparse
-import runpy
-import sys
-from pathlib import Path
 
-import torchtune
-from torch.distributed.run import get_args_parser, run
-from torchtune import list_recipes
-from torchtune._cli import list_scripts
-from torchtune.utils._distributed import _valid_distributed_single_node_nnodes
+from torchtune._cli.cp import Copy
+from torchtune._cli.download import Download
+from torchtune._cli.ls import List
+from torchtune._cli.run import Run
+from torchtune._cli.validate import Validate
 
 
-def _update_parser_help(parser):
-    parser.description = "Torch Tune Recipe Launcher"
-    parser.usage = "tune [options] <recipe> [recipe_args]"
-    parser.formatter_class = argparse.RawDescriptionHelpFormatter
+class TuneCLIParser:
+    """Holds all information related to running the CLI"""
 
-    # Update torchrun argparse name for more accurate CLI help
-    actions = [a.dest for a in parser._actions]
-    # Update training_script help to be recipe
-    idx = actions.index("training_script")
-    parser._actions[idx].dest = "recipe"
-    parser._actions[idx].help = "Name or path to recipe to be launched followed by args"
-
-    # Update training_script_args help to be recipe_args
-    idx = actions.index("training_script_args")
-    parser._actions[idx].dest = "recipe_args"
-
-
-def _is_distributed_args(args):
-    total = len(sys.argv) - 1  # total args minus "tune"
-    script_args = len(args.recipe_args) + 1  # script args + 1 for script name
-    return total > script_args
-
-
-def _validate_distributed_args(args):
-    """
-    Validates nnodes and nproc_per_node are appropriately set for distributed training
-    runs.
-    """
-    if not hasattr(args, "nnodes"):
-        raise RuntimeError("Expect --nnodes to be specified for distributed runs")
-
-    if args.nnodes not in _valid_distributed_single_node_nnodes:
-        raise RuntimeError(
-            f"Expect --nnodes to be one of {_valid_distributed_single_node_nnodes}"
+    def __init__(self):
+        # Initialize the top-level parser
+        self._parser = argparse.ArgumentParser(
+            prog="tune",
+            description="Welcome to the TorchTune CLI!",
+            add_help=True,
         )
+        # Default command is to print help
+        self._parser.set_defaults(func=lambda args: self._parser.print_help())
 
-    if not hasattr(args, "nproc_per_node"):
-        raise RuntimeError(
-            "Expect --nproc_per_node to be specified for distributed runs"
-        )
+        # Add subcommands
+        subparsers = self._parser.add_subparsers(title="subcommands")
+        Download.create(subparsers)
+        List.create(subparsers)
+        Copy.create(subparsers)
+        Run.create(subparsers)
+        Validate.create(subparsers)
```

Comment:
so these are class methods, which will create the subcommand instances for you... but then how does `tune` have access to the `self._parser` in each of these if you're not assigning the created instance to anything?

Reply:
Yeah, I hate this too. I wish there was a way in which I could add a full subparser to a top level argparse, but that's not how they add_subparser API is designed. I finally gave up on trying to find a way around this when I saw that accelerate adopted this approach, too. https://github.com/huggingface/accelerate/blob/main/src/accelerate/commands/accelerate_cli.py